python/cudf/cudf/core/frame.py

```diff
@@ -438,128 +438,241 @@ def _normalize_columns_and_scalars_type(self, other):
+                            categories=self._data[column_name].categories,
+                            codes=as_column(
+                                result.base_data, dtype=result.dtype
+                            ),
+                            mask=result.base_mask,
+                            size=result.size,
+                            offset=result.offset,
+                            ordered=self._data[column_name].ordered,
+                        )
+                else:
+                    from cudf._lib.null_mask import MaskState, create_null_mask
 
-                if is_categorical_dtype(self[in_col_name].dtype):
-                    result = build_categorical_column(
-                        categories=self[in_col_name]._column.categories,
-                        codes=as_column(result.base_data, dtype=result.dtype),
-                        mask=result.base_mask,
-                        size=result.size,
-                        offset=result.offset,
-                        ordered=self[in_col_name]._column.ordered,
+                    out_mask = create_null_mask(
+                        len(input_col), state=MaskState.ALL_NULL
                     )
+                    result = input_col.set_mask(out_mask)
+                out_df[column_name] = self[column_name].__class__(result)
 
-                out_df[in_col_name] = self[in_col_name].__class__(result)
-
-            return out_df
+            return self._mimic_inplace(out_df, inplace=inplace)
 
         else:
+            if isinstance(self, cudf.core.multiindex.MultiIndex):
+                raise NotImplementedError(
+                    ".where is not supported for MultiIndex operations"
+                )
+
+            if isinstance(other, cudf.DataFrame):
+                raise NotImplementedError(
+                    "cannot align with a higher dimensional Frame"
+                )
+
             other = self._normalize_columns_and_scalars_type(other)
-            if isinstance(other, (cudf.Series, cudf.Index)):
-                other = other._column
 
-            input_col = self._column
+            cond = as_column(cond)
+            if len(cond) != len(self):
+                raise ValueError(
+                    """Array conditional must be same shape as self"""
+                )
+            input_col = self._data[self.name]
             if is_categorical_dtype(input_col.dtype):
-                if np.isscalar(otr_col):
-                    other = input_col._encode(other)
-                else:
+                if np.isscalar(other):
+                    try:
+                        other = input_col._encode(other)
+                    except ValueError:
+                        # When other is not present in categories.
+                        other = input_col.default_na_value()
+                elif hasattr(other, "codes"):
                     other = other.codes
+
                 input_col = input_col.codes
```

Comment:
What if `input_col` isn't a `CategoricalColumn`?

Reply:
This is shielded inside a `is_categorical_dtype(input_col.dtype)` check.